### PR TITLE
Update kpm list

### PR DESCRIPTION
**Pull request by troydai**
1. Group the output, print framework dependencies first:
image
2. Add option to hide dependent. Produce compact output, improve readability.
image
3. Add additional empty line before and after every group of target framework, also intend to improve readability.
image
4. Add option to filter the result using wildcard pattern:
image
5. Fix issues:
https://github.com/aspnet/XRE/issues/1096
https://github.com/aspnet/XRE/issues/1094
https://github.com/aspnet/XRE/issues/1098

### CODE REVIEW

**Comment by troydai** [comment]
@davidfowl @Praburaj

File: src/Microsoft.Framework.PackageManager/List/LibraryDependencyFlatRenderer.cs [hunk 1]
```diff
@@ -10,21 +10,21 @@ namespace Microsoft.Framework.PackageManager.List
...
         public void Render(IGraphNode<Library> root)
         {
             var dict = FindImmediateDependent(root);
-            foreach (var key in dict.Keys.OrderBy(library => library.Name))
-            {
-                _output.WriteLine(key.ToString().White().Bold());
-                _output.WriteLine("    -> {0}", string.Join(", ", dict[key].Select(lib => lib.ToString()).OrderBy(s => s)));
-            }
+
+            RenderLibraries(dict.Keys.Where(library => library.IsGacOrFrameworkReference), dict);
```
Comment (Praburaj) [inline src/Microsoft.Framework.PackageManager/List/LibraryDependencyFlatRenderer.cs]: Can we optimize this a bit by storing the OrderedKeys on a local variable and do the where check - RenderLibraries can remove OrderBy as well.

File: src/Microsoft.Framework.PackageManager/List/LibraryDependencyFlatRenderer.cs [hunk 2]
```diff
@@ -10,21 +10,21 @@ namespace Microsoft.Framework.PackageManager.List
 {
     public class LibraryDependencyFlatRenderer
     {
-        private IReport _output;
+        private readonly IReport _output;
+        private readonly bool _hideDependent;
```
Comment (Praburaj) [inline src/Microsoft.Framework.PackageManager/List/LibraryDependencyFlatRenderer.cs]: can we rename this to `_hideDependencies`?

Comment (Praburaj) [inline src/Microsoft.Framework.PackageManager/List/LibraryDependencyFlatRenderer.cs]: alternatively `_listDependencies` and give a default value in the constructor.

File: src/Microsoft.Framework.PackageManager/Program.cs [hunk 3]
```diff
@@ -414,6 +414,9 @@ public int Main(string[] args)
                 var runtimeFolder = c.Option("--runtime",
                     "The folder containing all available framework assemblies.",
                     CommandOptionType.SingleValue);
+                var hideDependent = c.Option("--hide-dependent",
+                    "Hide the dependents",
```
Comment (Praburaj) [inline src/Microsoft.Framework.PackageManager/Program.cs]: `dependents` => `dependencies`.

Comment (troydai) [inline src/Microsoft.Framework.PackageManager/Program.cs]: It's not the dependencies, it's indeed __dependent__ which introduce the underline packages.

**Comment by troydai** [comment]
@Praburaj @davidfowl @muratg :watch:

File: src/Microsoft.Framework.PackageManager/List/DependencyListCommand.cs [hunk 4]
```diff
@@ -39,7 +38,7 @@ public int Execute()
 
             foreach (var framework in frameworks)
             {
-                _options.Reports.Information.WriteLine("[Target framework {0}]", framework.Identifier.ToString());
+                _options.Reports.Information.WriteLine("\n[Target framework {0}]\n", framework.Identifier.ToString());
```
Comment (davidfowl) [inline src/Microsoft.Framework.PackageManager/List/DependencyListCommand.cs]: Use framework, not the identifier

File: src/Microsoft.Framework.PackageManager/List/DependencyListOptions.cs [hunk 5]
```diff
@@ -59,6 +59,10 @@ public DependencyListOptions(Reports reports, CommandArgument path, CommandOptio
 
         public FrameworkName Framework { get; }
 
+        public bool HideDependent { get; set; }
```
Comment (davidfowl) [inline src/Microsoft.Framework.PackageManager/List/DependencyListOptions.cs]: HideDependents?

File: src/Microsoft.Framework.PackageManager/Program.cs [hunk 6]
```diff
@@ -404,17 +404,23 @@ public int Main(string[] args)
...
+                var framework = c.Option("--framework <NAME>",
+                    "Show dependencies for only the given framework",
                     CommandOptionType.SingleValue);
-                var runtimeFolder = c.Option("--runtime",
-                    "The folder containing all available framework assemblies.",
+                var runtimeFolder = c.Option("--runtime <PATH>",
+                    "The folder containing all available framework assemblies",
                     CommandOptionType.SingleValue);
-                var argProject = c.Argument("[project]", "The path to project. If omitted, the command will use the project in the current directory.");
+                var hideDependent = c.Option("--hide-dependent",
```
Comment (davidfowl) [inline src/Microsoft.Framework.PackageManager/Program.cs]: --hide-dependents

File: src/Microsoft.Framework.PackageManager/List/DependencyListOptions.cs (hunk 5 above)
Comment (troydai) [inline src/Microsoft.Framework.PackageManager/List/DependencyListOptions.cs]: ok

File: src/Microsoft.Framework.PackageManager/Program.cs [hunk 7]
```diff
@@ -404,17 +404,23 @@ public int Main(string[] args)
...
+                    "Hide the dependents",
```
Comment (davidfowl) [inline src/Microsoft.Framework.PackageManager/Program.cs]: That's a poor description

**Comment by troydai** [comment]
:watch:

File: src/Microsoft.Framework.PackageManager/List/DependencyListCommand.cs [hunk 8]
```diff
@@ -39,7 +39,7 @@ public int Execute()
 
             foreach (var framework in frameworks)
             {
-                _options.Reports.Information.WriteLine("[Target framework {0}]", framework.Identifier.ToString());
+                _options.Reports.Information.WriteLine("\n[Target framework {0} ({1})]\n", framework.Identifier.ToString(), VersionUtility.GetShortFrameworkName(framework));
```
Comment (davidfowl) [inline src/Microsoft.Framework.PackageManager/List/DependencyListCommand.cs]: Use the framework, not just the identifier

File: src/Microsoft.Framework.PackageManager/List/LibraryDependencyFlatRenderer.cs (hunk 2 above)
Comment (davidfowl) [inline src/Microsoft.Framework.PackageManager/List/LibraryDependencyFlatRenderer.cs]: hideDependents

File: src/Microsoft.Framework.PackageManager/List/LibraryDependencyFlatRenderer.cs [hunk 9]
```diff
@@ -50,7 +54,34 @@ public void Render(IGraphNode<Library> root)
...
 
+            // removing the root package
+            result.Remove(root.Item);
+
             return result;
         }
+
+        private void RenderLibraries(IEnumerable<Library> libraries, IDictionary<Library, ISet<Library>> dependenciesMap)
+        {
+            if (_filterPattern != null)
```
Comment (davidfowl) [inline src/Microsoft.Framework.PackageManager/List/LibraryDependencyFlatRenderer.cs]: string.IsNullOrEmpty

File: src/Microsoft.Framework.PackageManager/Program.cs [hunk 10]
```diff
@@ -404,17 +404,23 @@ public int Main(string[] args)
...
-                    "The folder containing all available framework assemblies.",
+                var runtimeFolder = c.Option("--runtime <PATH>",
+                    "The folder containing all available framework assemblies",
                     CommandOptionType.SingleValue);
-                var argProject = c.Argument("[project]", "The path to project. If omitted, the command will use the project in the current directory.");
+                var hideDependents = c.Option("--hide-dependents",
+                    "Hide the immediate dependents of libraries referenced in the project",
+                    CommandOptionType.NoValue);
+                var resultsFilter = c.Option("--filter <PATTERN>",
+                    "Filter the result libraries base on their names. The matching pattern supports * and ?",
```
Comment (davidfowl) [inline src/Microsoft.Framework.PackageManager/Program.cs]: Result libraries doesn't look nice. Sick to `libraries referenced by the project`

**Comment by troydai** [comment]
@davidfowl looks good?

**Pull request merged**